device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java

```diff
@@ -1336,6 +1022,207 @@ public void onSubscriptionMessageSent(int deliveryTag, SubscriptionType subscrip
         this.inProgressSubscriptionMessages.put(deliveryTag, subscriptionType);
     }
 
+    @Override
+    public void onDeviceSessionOpened(String deviceId)
+    {
+        this.deviceSessionsOpenedLatch.countDown();
+    }
+
+    @Override
+    public void onAuthenticationSessionOpened() {
+        this.authenticationSessionOpenedLatch.countDown();
+    }
+
+    final void addDeviceSession(DeviceClientConfig deviceClientConfig, SubscriptionMessageRequestSentCallback subscriptionMessageRequestSentCallback, boolean afterOpen) throws TransportException {
+        if (deviceClientConfig == null)
+        {
+            throw new IllegalArgumentException("deviceClientConfig cannot be null.");
+        }
+
+        // Check if the device session still exists from a previous connection
+        AmqpsSessionHandler amqpsSessionHandler = null;
+        for (AmqpsSessionHandler existingAmqpsSessionHandler : this.sessionHandlerList)
+        {
+            if (existingAmqpsSessionHandler.getDeviceId().equals(deviceClientConfig.getDeviceId()))
+            {
+                amqpsSessionHandler = existingAmqpsSessionHandler;
+                break;
+            }
+        }
+
+        // If the device session did not exist in the previous connection, or if there was no previous connection,
+        // create a new session
+        if (amqpsSessionHandler == null)
+        {
+            amqpsSessionHandler = new AmqpsSessionHandler(deviceClientConfig, subscriptionMessageRequestSentCallback);
+            this.sessionHandlerList.add(amqpsSessionHandler);
+        }
+
+        if (afterOpen)
+        {
+            amqpsSessionHandler.session = this.connection.session();
+            amqpsSessionHandler.session.open();
+
+            this.authenticate(amqpsSessionHandler);
+        }
+    }
+
+    public void authenticate(AmqpsSessionHandler sessionHandler) throws TransportException
+    {
+        if (!sessionHandler.isAuthenticated())
+        {
+            UUID authenticationCorrelationId = UUID.randomUUID();
+            sessionHandler.cbsCorrelationIdList.add(authenticationCorrelationId);
+            log.debug("Sending authentication message with correlationId {} to authenticate device {}", authenticationCorrelationId, sessionHandler.getDeviceId());
+            this.amqpsAuthenticationLinkHandler.authenticate(sessionHandler.deviceClientConfig, authenticationCorrelationId);
+        }
+    }
+
+    public void authenticate() throws TransportException
+    {
+        for (AmqpsSessionHandler sessionHandler : this.sessionHandlerList)
+        {
+            authenticate(sessionHandler);
```

Comment:
authenticate for each session, why? multiplexed sessions will still have a single cbs auth link, right?

Reply:
Each session uses the same CBS links, but each session has to send a message over those links to authenticate its corresponding device. This method just says to authenticate each device over the open CBS links.